src/value-parser.ts

```diff
@@ -14,702 +15,757 @@ const PLP_NULL = Buffer.from([0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF]);
-  });
+  return new Result(iconv.decode(buf.slice(offset, offset + dataLength), codepage ?? DEFAULT_ENCODING), offset + dataLength);
 }
 
-function readMaxNChars(parser: Parser, callback: (value: string | null) => void) {
-  readMax(parser, (data) => {
-    if (data) {
-      callback(data.toString('ucs2'));
-    } else {
-      callback(null);
-    }
-  });
+function readNChars(buf: Buffer, offset: number, dataLength: number): Result<string> {
+  if (buf.length < offset + dataLength) {
+    throw new NotEnoughDataError(offset + dataLength);
+  }
+
+  return new Result(buf.toString('ucs2', offset, offset + dataLength), offset + dataLength);
 }
 
-function readMax(parser: Parser, callback: (value: null | Buffer) => void) {
-  parser.readBuffer(8, (type) => {
-    if (type.equals(PLP_NULL)) {
-      return callback(null);
-    } else if (type.equals(UNKNOWN_PLP_LEN)) {
-      return readMaxUnknownLength(parser, callback);
-    } else {
-      const low = type.readUInt32LE(0);
-      const high = type.readUInt32LE(4);
-
-      if (high >= (2 << (53 - 32))) {
-        console.warn('Read UInt64LE > 53 bits : high=' + high + ', low=' + low);
-      }
+async function readPLPStream(parser: Parser): Promise<null | Buffer[]> {
+  while (parser.buffer.length < parser.position + 8) {
+    await parser.waitForChunk();
+  }
 
-      const expectedLength = low + (0x100000000 * high);
-      return readMaxKnownLength(parser, expectedLength, callback);
+  const type = parser.buffer.slice(parser.position, parser.position + 8);
+  parser.position += 8;
+
+  if (type.equals(PLP_NULL)) {
+    return null;
+  }
+
+  const chunks: Buffer[] = [];
+  let currentLength = 0;
+
+  while (true) {
+    while (parser.buffer.length < parser.position + 4) {
+      await parser.waitForChunk();
     }
-  });
-}
 
-function readMaxKnownLength(parser: Parser, totalLength: number, callback: (value: null | Buffer) => void) {
-  const data = Buffer.alloc(totalLength, 0);
+    const chunkLength = parser.buffer.readUInt32LE(parser.position);
+    parser.position += 4;
 
-  let offset = 0;
-  function next(done: any) {
-    parser.readUInt32LE((chunkLength) => {
-      if (!chunkLength) {
-        return done();
-      }
+    if (!chunkLength) {
+      break;
+    }
 
-      parser.readBuffer(chunkLength, (chunk) => {
-        chunk.copy(data, offset);
-        offset += chunkLength;
+    while (parser.buffer.length < parser.position + chunkLength) {
+      await parser.waitForChunk();
+    }
 
-        next(done);
-      });
-    });
+    chunks.push(parser.buffer.slice(parser.position, parser.position + chunkLength));
+    parser.position += chunkLength;
+    currentLength += chunkLength;
   }
 
-  next(() => {
-    if (offset !== totalLength) {
-      throw new Error('Partially Length-prefixed Bytes unmatched lengths : expected ' + totalLength + ', but got ' + offset + ' bytes');
+  if (!type.equals(UNKNOWN_PLP_LEN)) {
+    const expectedLength = Number(readBigInt(type, 0).value);
```

Comment:
The `type` is guaranteed to be a 8 byte `Buffer` here - I think we can call `type.readBigUInt64LE` directly on it, without having to go through any of the helper functions.